File: locus-api-android/src/main/java/locus/api/android/objects/ParcelableContainer.java
package locus.api.android.objects;

import android.os.Parcel;
import android.os.Parcelable;

/**
 * Created by menion on 10. 7. 2014.
 * Class is part of Locus project
 */
public class ParcelableContainer implements Parcelable {

    // stored data container
    private byte[] mData;

    public ParcelableContainer(byte[] data) {
        // check data
        if (data == null) {
            throw new IllegalArgumentException("'data' cannot 'null'");
        }

        // store data container
        this.mData = data;
    }

    private ParcelableContainer(Parcel in) {
        readFromParcel(in);
    }

    /**
     * Return inner byte array
     * @return
     */
    public byte[] getData() {
        return mData;
    }

    // PARCELABLE PART

    public static final Parcelable.Creator<ParcelableContainer> CREATOR = new
            Parcelable.Creator<ParcelableContainer>() {

                public ParcelableContainer createFromParcel(Parcel in) {
                    return new ParcelableContainer(in);
                }

                public ParcelableContainer[] newArray(int size) {
                    return new ParcelableContainer[size];
                }
            };

    @Override
    public int describeContents() {
        return 0;
    }

    private void readFromParcel(Parcel in) {
        mData = new byte[in.readInt()];
        in.readByteArray(mData);
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        // write data type
        dest.writeInt(mData.length);
        dest.writeByteArray(mData);
    }
}

File: locus-api-android/src/main/java/locus/api/android/objects/ParcelableContainer.kt
/**
 * Created by menion on 10. 7. 2014.
 * Class is part of Locus project
 */
package locus.api.android.objects

import android.os.Parcel
import android.os.Parcelable

/**
 * Nice parcelable implementation that is able to carry byte array between consumers.
 */
class ParcelableContainer : Parcelable {

    /**
     * Inner byte array
     */
    var data: ByteArray? = null
        private set

    constructor(data: ByteArray?) {
        // check data
        if (data == null) {
            throw IllegalArgumentException("'data' cannot 'null'")
        }

        // store data container
        this.data = data
    }

    private constructor(`in`: Parcel) {
        readFromParcel(`in`)
    }

    override fun describeContents(): Int {
        return 0
    }

    private fun readFromParcel(`in`: Parcel) {
        data = `in`.readSizedByteArray()
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        // write data type
        dest.writeInt(data!!.size)
        dest.writeByteArray(data)
    }

    companion object CREATOR : Parcelable.Creator<ParcelableContainer> {
        override fun createFromParcel(parcel: Parcel): ParcelableContainer {
            return ParcelableContainer(parcel)
        }

        override fun newArray(size: Int): Array<ParcelableContainer?> {
            return arrayOfNulls(size)
        }
    }
}

/**
 * Read a byte array written as its size followed by `Parcel.writeByteArray()`.
 * The array is bounded by the parcel content, never allocated from the size alone.
 */
internal fun Parcel.readSizedByteArray(): ByteArray {
    val size = readInt()
    val data = createByteArray()
    if (data == null || data.size != size) {
        throw RuntimeException("bad array lengths")
    }
    return data
}
